### ISSUE TITLE
sql: CHAR behavior inconsistent with Postgres

### ISSUE THREAD

**Issue by mgartner** [cockroachdb/cockroach#69673]
In Postgres:

```
marcus=# CREATE TABLE t (c CHAR);
CREATE TABLE

marcus=# INSERT INTO t VALUES (' '), ('');
INSERT 0 2

marcus=# SELECT c = ' ', c = '' FROM t;
 ?column? | ?column?
----------+----------
 t        | t
 t        | t
(2 rows)
```

In cockroach v21.1.7:

```
defaultdb> CREATE TABLE t (c CHAR);
CREATE TABLE

defaultdb> INSERT INTO t VALUES (' '), ('');
INSERT 2

defaultdb> SELECT c = ' ', c = '' FROM t;
  ?column? | ?column?
-----------+-----------
   false   |   true
   false   |   true
(2 rows)
```

Jira issue: CRDB-9714

**Comment by rafiss** [comment]
More fun behavior on this:
```
rafiss@127:postgres> select length(c) from t where c = '';
+--------+
| length |
|--------|
| 0      |
| 0      |
+--------+
SELECT 2
Time: 0.008s
rafiss@127:postgres> select length(c) from t where c = ' ';
+--------+
| length |
|--------|
| 0      |
| 0      |
+--------+
```

**Comment by mgartner** [comment]
This was fixed by #133037. The fix might be disabled by default—see #137844 and the `legacy_varchar_typing` session setting.


On master@1fe09ec0f:

```sql
CREATE TABLE t (c CHAR);

INSERT INTO t VALUES (' '), ('');

SELECT c = ' ', c = '' FROM t;
--   ?column? | ?column?
-- -----------+-----------
--      t     |    t
--      t     |    t
-- (2 rows)

SELECT length(c) FROM t WHERE c = '';
--   length
-- ----------
--        0
--        0
-- (2 rows)

SELECT length(c) FROM t WHERE c = ' ';
--   length
-- ----------
--        0
--        0
-- (2 rows)
```
